### PR TITLE
feat(subscription): add support for time interval #517

### PR DESCRIPTION
**Pull request by bobeal**

### CODE REVIEW

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt [hunk 1]
```diff
@@ -64,20 +71,26 @@ class SubscriptionHandler(
...
-        )
 
-        return ResponseEntity.status(HttpStatus.CREATED)
-            .location(URI("/ngsi-ld/v1/subscriptions/${parsedSubscription.id}"))
-            .build<String>()
+        return either<APiException, ResponseEntity<*>>{
+            val parsedSubscription = parseSubscription(body, contexts)
+            checkSubscriptionNotExists(parsedSubscription).awaitFirst().bind()
+
+            subscriptionService.create(parsedSubscription, sub)
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt:80]: `subscription.create(...)` returns a `Mono`. If you don't subscribe to it in some way, it will never be executed.

Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt:80]: `subscription.create(...)` returns a `Mono`. If you don't subscribe to it in some way, it will never be executed.

**Comment by GaelPoujolSenseen** [comment]
I have read the CLA Document and I hereby sign the CLA

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/utils/ParsingUtils.kt [hunk 2]
```diff
@@ -80,4 +89,17 @@ object ParsingUtils {
...
+
+    fun checkConflitAttribute(subscription: Subscription): Either<APiException, Subscription> =
+        if (subscription.watchedAttributes != null && subscription.timeInterval != null)
+            BadRequestDataException(
+                "You can't use 'timeInterval' with 'watchedAttributes' mustn't be used in conjunction"
+            )
+                .left()
+        else subscription.right()
+
+    fun checkTimeIntervalNotEmpty(subscription: Subscription): Either<APiException, Subscription> =
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/utils/ParsingUtils.kt:101]: "NotEmpty" is not exactly what you are checking

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt [hunk 3]
```diff
@@ -231,6 +231,63 @@ class SubscriptionHandlerTests {
...
+                    "type":"https://uri.etsi.org/ngsi-ld/errors/BadRequestData",
+                    "title":"The request includes input data which does not meet the requirements of the operation",
+                    "detail":"The value of 'timeInterval' must be greater than zero (int)"
+                } 
+                """.trimIndent()
+            )
+    }
+
+    @Test
+    fun `create subscription should return a 201 if JSON-LD payload with 'timeInterval' is correct`() {
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt:277]: not sure this one is really needed

File: subscription-service/src/test/resources/ngsild/subscription_with_time_interval_less_than_0.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    },
+    { "id": "urn:ngsi-ld:Beehive:1234567890",
+      "type": "Beehive"
+    },
+    {
+      "idPattern": "urn:ngsi-ld:Beehive:1234*",
+      "type": "Beehive"
+    }
+  ],
+  "timeInterval": 0.1,
```
Comment (bobeal) [inline subscription-service/src/test/resources/ngsild/subscription_with_time_interval_less_than_0.json:16]: "0.1" is not "less than 0" :)

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt [hunk 5]
```diff
@@ -219,7 +219,7 @@ class SubscriptionHandlerTests {
             .uri("/ngsi-ld/v1/subscriptions")
             .bodyValue(jsonLdFile)
             .exchange()
-            .expectStatus().isBadRequest
+            .expectStatus().is4xxClientError
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt:222]: I had changed it for a test and I forgot to modify it

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt (hunk 3 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt:277]: ok I take it off

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt [hunk 6]
```diff
@@ -231,6 +231,63 @@ class SubscriptionHandlerTests {
...
+                    "type":"https://uri.etsi.org/ngsi-ld/errors/BadRequestData",
+                    "title":"The request includes input data which does not meet the requirements of the operation",
+                    "detail":"You can't use 'timeInterval' with 'watchedAttributes' mustn't be used in conjunction"
+                } 
+                """.trimIndent()
+            )
+    }
+
+    @Test
+    fun `create subscription should return a 400 if JSON-LD payload contains 'timeInterval' less than 0 '`() {
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/web/SubscriptionHandlerTests.kt:256]: Maybe 'create subscription should return a 400 if JSON-LD payload contains 'timeInterval' less than 1' like 0 can't be used too

File: subscription-service/src/test/resources/ngsild/subscription_with_time_interval_less_than_0.json (hunk 4 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/resources/ngsild/subscription_with_time_interval_less_than_0.json:16]: Yes I admit that the name is not good for this use case. "les than 1" is better

File: subscription-service/src/test/resources/ngsild/subscription_with_time_interval.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (bobeal) [inline subscription-service/src/test/resources/ngsild/subscription_with_time_interval.json:1]: don't forget to remove this file, you're no longer using it

File: subscription-service/src/test/resources/ngsild/subscription_with_conflit_attribute.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (bobeal) [inline subscription-service/src/test/resources/ngsild/subscription_with_conflit_attribute.json:1]: try to find a better name for this file. one that indicates why it is not a correct subscription.

Comment (GaelPoujolSenseen) [reply subscription-service/src/test/resources/ngsild/subscription_with_conflit_attribute.json:1]: Maybe 'subscription_with_invalid_attribut' ?

File: subscription-service/src/test/resources/ngsild/subscription_with_time_interval.json (hunk 7 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/resources/ngsild/subscription_with_time_interval.json:1]: I do it

File: subscription-service/src/test/resources/ngsild/subscription_with_conflit_attribute.json (hunk 8 above)
Comment (bobeal) [reply subscription-service/src/test/resources/ngsild/subscription_with_conflit_attribute.json:1]: still too vague, we can have many cases of invalid subscriptions... `subscription_with_conflicting_timeInterval_watchedAttributes`?

Comment (GaelPoujolSenseen) [reply subscription-service/src/test/resources/ngsild/subscription_with_conflit_attribute.json:1]: I valid

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+class ScheduledNotification(
+    private val subscriptionService: SubscriptionService,
+    private val notificationService: NotificationService
+) {
+
+    @Autowired
+    private val webClient: WebClient? = null
+
+    /**
+     * Every seconds, the function cheks the notification to send
```
[the remark this answers is not in the document]
Comment (bobeal) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:30]: and globally same for almost all comments in this class: add one if it brings something that is not evident from the signature of the method.

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            entity.id.toUri(),
+            entity
+        )
+
+    /**
+     * We call entity service and we want a list of entity
+     *
+     * @return list of JsonLdEntity
+     */
+    suspend fun getEntitiesByType(entityId: String): List<JsonLdEntity> =
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:102]: the method parameter is not an id of an entity

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        webClient!!.get()
+            .uri("/entities$entityId")
+            .attributes(clientRegistrationId("keycloak"))
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:105]: do you need to specify this, since it is the default?

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        return listResultSendNotification
+    }
+
+    /**
+     * Function to send notificatio
+     * We call 'callSubscriber'
+     *
+     * @return Mono<Triple<Subscription, Notification, Boolean>>
+     */
+    fun sendNotification(entity: JsonLdEntity, subscription: Subscription):
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:89]: if you don't subscribe to the result (or use one the `awaitXXX` function in the calling code), the `callSubscriber(...)` will never be executed since it returns a Mono (like a promise of a result)

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/WebClientConfig.kt [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.security.oauth2.client.registration.ClientRegistration
+import org.springframework.security.oauth2.client.registration.InMemoryReactiveClientRegistrationRepository
+import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository
+import org.springframework.security.oauth2.client.web.reactive.function.client.ServerOAuth2AuthorizedClientExchangeFilterFunction
+import org.springframework.security.oauth2.core.AuthorizationGrantType
+import org.springframework.web.reactive.function.client.WebClient
+
+@PropertySource("classpath:application.properties")
+@Configuration
+class WebClientConfig {
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/WebClientConfig.kt:18]: why don't you have a simple config like in https://www.baeldung.com/spring-webclient-oauth2#spring-security-5-support---the-client-credentials-flow?

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt [hunk 14]
```diff
@@ -591,4 +586,42 @@ class SubscriptionService(
...
         row.get("sub", String::class.java)!!
     }
+
+    /**
+     * Function that returns all subscriptions that contain the 'timeInterval' attribute and
+     * when the 'lastNotification' has passed
+     *
+     * @return list of subscription
+     */
+    suspend fun getAllSubscriptions(): List<Subscription> {
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:596]: the name of this function is very misleading.

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt [hunk 15]
```diff
@@ -591,4 +586,42 @@ class SubscriptionService(
...
+                description, watched_attributes, time_interval, q,  notif_attributes, notif_format, endpoint_uri, 
+                endpoint_accept, endpoint_info,  status, times_sent, last_notification, last_failure, last_success, 
+                is_active, entity_info.id as entity_id, id_pattern, entity_info.type as entity_type, georel, geometry, 
+                coordinates, geoproperty
+            FROM subscription
+            LEFT JOIN entity_info ON entity_info.subscription_id = subscription.id
+            LEFT JOIN geometry_query ON geometry_query.subscription_id = subscription.id
+            WHERE time_interval IS NOT NULL
+            AND last_notification IS NULL 
+            OR ((EXTRACT(EPOCH FROM last_notification) + time_interval) < EXTRACT(EPOCH FROM :currentDate))
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:609]: - don't you miss parenthesis to group these 2 conditions?
- should not the condition between the 2 dates be in the opposite way?

File: subscription-service/src/main/resources/application.properties [hunk 16]
```diff
@@ -30,7 +30,12 @@ application.authentication.enabled = false
...
-
 # Pagination config for query resources endpoints
-application.pagination.limit-default = 30
-application.pagination.limit-max = 100
+application.pagination.limit-default=30
+application.pagination.limit-max=100
+#configure ressource for authentification
+spring.security.oauth2.client.registration.keycloak.authorization-grant-type=client_credentials
+spring.security.oauth2.client.registration.keycloak.client-id=senseen-gael
+spring.security.oauth2.client.registration.keycloak.client-secret=221d117c-b7fd-4b8d-8064-f9c51605c81c
```
Comment (bobeal) [inline subscription-service/src/main/resources/application.properties:39]: this is very bad! (btw, I removed all your sessions and changed your secret...)

File: subscription-service/src/main/resources/application.properties [hunk 17]
```diff
@@ -30,7 +30,12 @@ application.authentication.enabled = false
...
+spring.security.oauth2.client.provider.keycloak.token-uri=https://sso.eglobalmark.com/auth/realms/stellio-dev/protocol/openid-connect/token
+spring.security.oauth2.client.provider.keycloak.base-url=https://stellio-dev.eglobalmark.com
```
Comment (bobeal) [inline subscription-service/src/main/resources/application.properties:41]: this is not related to KC (check how similar property is named in search service)

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    private lateinit var notificationService: NotificationService
+
+    @MockkBean
+    private lateinit var subscriptionService: SubscriptionService
+
+    @MockkBean
+    private lateinit var subscriptionEventService: SubscriptionEventService
+
+    @MockkBean
+    private lateinit var fcmService: FCMService
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt:50]: you are not using these 2 services in your job, they should not appear here.

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+     */
+    @Test
+    fun `it should return one entity`() {
+        WireMock.stubFor(
+            WireMock.get(WireMock.urlMatching("/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:TESTC"))
+                .willReturn(WireMock.okJson(loadSampleData("beehive.jsonld")))
+        )
+
+        runBlocking {
+            val entity = scheduledNotification.getEntity("/urn:ngsi-ld:BeeHive:TESTC")
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt:137]: a bit strange to have this leading `/` for entity id

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 20]
```diff
@@ -899,4 +899,90 @@ class SubscriptionServiceTests : WithTimescaleContainer {
...
+
+        StepVerifier.create(updateResult)
+            .expectNextMatches {
+                it.id == subscription1Id &&
+                    it.notification.status == StatusType.OK &&
+                    it.notification.timesSent == 1 &&
+                    it.notification.lastNotification != null &&
+                    it.notification.lastSuccess != null &&
+                    it.notification.lastFailure == null
+            }
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:928]: you don't really care about this (I guess this is already tested somewhere in this class). what you want to test here is that you get the expected subscriptions:
- never notified and time interval reached compared to creation time of the subscription
- already notified and time interval reached since the last notification

File: subscription-service/src/test/resources/application-test.properties [hunk 21]
```diff
@@ -1 +1,7 @@
-application.authentication.enabled = true
+application.authentication.enabled=true
+#configure ressource for authentification
+spring.security.oauth2.client.registration.keycloak.authorization-grant-type=client_credentials
+spring.security.oauth2.client.registration.keycloak.client-id=senseen-gael
+spring.security.oauth2.client.registration.keycloak.client-secret=221d117c-b7fd-4b8d-8064-f9c51605c81c
+spring.security.oauth2.client.provider.keycloak.token-uri=https://sso.eglobalmark.com/auth/realms/stellio-dev/protocol/openid-connect/token
+spring.security.oauth2.client.provider.keycloak.base-url=http://localhost:8089/ngsi-ld/v1
```
Comment (bobeal) [inline subscription-service/src/test/resources/application-test.properties:7]: same as for `application.properties`

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt [hunk 22]
```diff
@@ -0,0 +1,125 @@
+package com.egm.stellio.subscription.config
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:1]: I do this

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+import reactor.core.publisher.Mono
+
+@Component
+class ScheduledNotification(
+    private val subscriptionService: SubscriptionService,
+    private val notificationService: NotificationService
+) {
+
+    @Autowired
+    private val webClient: WebClient? = null
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:27]: It's changed

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:30]: Ok I delete them all

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt (hunk 12 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:89]: I add awaitFirst

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt (hunk 14 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:596]: maybe `getSubscriptionWithIntervalPast`

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt (hunk 15 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:609]: 1. Yes I forgot a parenthesis
2. If the condition are in opposite way, It's going to return me notification when the TimeInterval are not past

File: subscription-service/src/main/resources/application.properties (hunk 16 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/resources/application.properties:39]: Can I create a KC account ?

File: subscription-service/src/main/resources/application.properties (hunk 17 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/resources/application.properties:41]: I change this

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/SubscriptionServiceApplication.kt [hunk 24]
```diff
@@ -4,7 +4,9 @@ import com.egm.stellio.subscription.config.ApplicationProperties
 import org.springframework.boot.autoconfigure.SpringBootApplication
 import org.springframework.boot.context.properties.EnableConfigurationProperties
 import org.springframework.boot.runApplication
+import org.springframework.scheduling.annotation.EnableScheduling
 
+@EnableScheduling
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/SubscriptionServiceApplication.kt:9]: ok I modified that

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt (hunk 22 above)
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:1]: It's done

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt (hunk 23 above)
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:27]: I changed that

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:30]: I deleted them all

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt (hunk 10 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/ScheduledNotification.kt:102]: I modified that

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/WebClientConfig.kt (hunk 13 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/WebClientConfig.kt:18]: I don't use the function UnAuthenticatedServerOAuth2AuthorizedClientRepository() because it is deprecated
So I use other method. I see another one this morning. I'm going to test it

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt (hunk 14 above)
Comment (bobeal) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:596]: or `getRecurringSubscriptionsToNotify`?

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt (hunk 18 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt:50]: Ok I take them off

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt (hunk 15 above)
Comment (bobeal) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:609]: indeed for 2.

File: subscription-service/src/main/resources/application.properties (hunk 16 above)
Comment (bobeal) [reply subscription-service/src/main/resources/application.properties:39]: the problem is that you commited a secret on a public repository. you can still use the same client, just don't publish this info in a public place.

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt (hunk 19 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/config/ScheduledNotificationTest.kt:137]: It's because when I extractParam I put `/` when I'ts `id` and `?` the other times

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt (hunk 20 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:928]: Ok I changed that

File: subscription-service/src/main/resources/application.properties (hunk 16 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/resources/application.properties:39]: I understood but how can I put the secret on the application ?

Comment (bobeal) [reply subscription-service/src/main/resources/application.properties:39]: Secrets are never stored in the source code. They are specific to each deployment and injected at runtime.

Comment (GaelPoujolSenseen) [reply subscription-service/src/main/resources/application.properties:39]: so I can take off this line `spring.security.oauth2.client.registration.keycloak.client-secret` ?

Comment (bobeal) [reply subscription-service/src/main/resources/application.properties:39]: leave the property (that also makes a sort of referential of supported properties) but give it a default dumb value (same for client-id btw)

Comment (GaelPoujolSenseen) [reply subscription-service/src/main/resources/application.properties:39]: Idem for token-uri ? It change to each deployment ?

**Comment by bobeal** [comment]
yes

From: GaelPOUJOL ***@***.***>
Date: Wednesday, 16 March 2022 at 11:07
To: stellio-hub/stellio-context-broker ***@***.***>
Cc: Benoit ORIHUELA (EGM) ***@***.***>, Author ***@***.***>
Subject: Re: [stellio-hub/stellio-context-broker] feat(subscription): add support for time interval #517 (PR #575)

@GaelPoujolSenseen commented on this pull request.

________________________________

In subscription-service/src/main/resources/application.properties<https://github.com/stellio-hub/stellio-context-broker/pull/575#discussion_r827831709>:

> +spring.security.oauth2.client.registration.keycloak.client-id=senseen-gael

+spring.security.oauth2.client.registration.keycloak.client-secret=221d117c-b7fd-4b8d-8064-f9c51605c81c

Idem for token-uri ? It change to each deployment ?

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt [hunk 25]
```diff
@@ -43,6 +36,7 @@ import org.springframework.web.bind.annotation.RequestMapping
 import org.springframework.web.bind.annotation.RequestParam
 import org.springframework.web.bind.annotation.RestController
 import reactor.core.publisher.Mono
+import reactor.kotlin.extra.bool.not
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt:39]: don't think you are using this

File: subscription-service/src/main/resources/application.properties [hunk 26]
```diff
@@ -30,7 +30,12 @@ application.authentication.enabled = false
...
-
 # Pagination config for query resources endpoints
-application.pagination.limit-default = 30
-application.pagination.limit-max = 100
+application.pagination.limit-default=30
+application.pagination.limit-max=100
+#configure ressource for authentification
+spring.security.oauth2.client.registration.keycloak.authorization-grant-type=client_credentials
+spring.security.oauth2.client.registration.keycloak.client-id=keycloak-id
+spring.security.oauth2.client.registration.keycloak.client-secret=keycloak-secret
```
Comment (bobeal) [inline subscription-service/src/main/resources/application.properties:39]: name the default value `client-id` and `client-secret` (this is what they are)

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/Job/TimeIntervalNotificationJobTest.kt [hunk 27]
```diff
@@ -0,0 +1,340 @@
+package com.egm.stellio.subscription.Job
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/Job/TimeIntervalNotificationJobTest.kt:1]: the package name must be in lower case

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 28]
```diff
@@ -899,4 +901,54 @@ class SubscriptionServiceTests : WithTimescaleContainer {
...
+
+    /**
+     * Test for function 'getAllSubscriptions' with two subscription.
+     * One with the 'lastNotification' attribute set to null and the other with the 'lastNotification' now and
+     * 'timeInterval' set to 1
+     *
+     * @return list of subscription
+     */
+    @Test
+    fun `it should return all Entity where 'timeInterval' are expired `() {
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:913]:
```suggestion
    fun `it should return all subscriptions whose 'timeInterval' is reached`() {
```

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 29]
```diff
@@ -899,4 +901,54 @@ class SubscriptionServiceTests : WithTimescaleContainer {
...
         assertEquals(res, true)
     }
+
+    /**
+     * Test for function 'getAllSubscriptions' with two subscription.
+     * One with the 'lastNotification' attribute set to null and the other with the 'lastNotification' now and
+     * 'timeInterval' set to 1
+     *
+     * @return list of subscription
+     */
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:911]: hmm...

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt [hunk 30]
```diff
@@ -899,4 +901,54 @@ class SubscriptionServiceTests : WithTimescaleContainer {
...
+    @Test
+    fun `it should return all Entity where 'timeInterval' are expired `() {
+
+        val persistedSubscription = subscriptionService.getById(subscription1Id).block()!!
+        val notification = Notification(subscriptionId = subscription1Id, data = emptyList())
+
+        runBlocking {
+            subscriptionService.updateSubscriptionNotification(persistedSubscription, notification, true)
+                .then(subscriptionService.getById(subscription1Id)).awaitFirst()
```
Comment (bobeal) [inline subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:920]: don't really get what you are trying to do here:
- do you really need to get the subscription by id?
- what is the purpose of updating the notification?

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+    fun extractParam(entityInfo: EntityInfo, q: String?): String {
+        val param = java.lang.StringBuilder()
+        param.append("?type=${entityInfo.type}")
+        if (entityInfo.id != null) param.append("&id=${entityInfo.id}")
+        if (entityInfo.idPattern != null) param.append("&idPattern=${entityInfo.idPattern}")
+        if (q != null) param.append("&q=$q")
+        return param.toString()
+    }
+
+    suspend fun sendNotification(listEntity: List<JsonLdEntity>, subscription: Subscription):
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt:47]: add a TODO here because you should actually send only one notification per subscription

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt (hunk 30 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:920]: 1. I delete that `then(subscriptionService.getById(subscription1Id))` because I don't need 
2. I update the notification to update `lastNotification` attribute

File: subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt (hunk 29 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/test/kotlin/com/egm/stellio/subscription/service/SubscriptionServiceTests.kt:911]: Maybe One with the 'lastNotification' attribute set to null and the other with the 'lastNotification' with instant date?

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt [hunk 32]
```diff
@@ -0,0 +1,80 @@
...
+            entity
+        ).awaitFirst()
+
+    suspend fun getEntitiesToNotify(listEntityInfo: Set<EntityInfo>, q: String?): List<JsonLdEntity> {
+        val listJsonLdEntity = ArrayList<JsonLdEntity>()
+        listEntityInfo.forEach {
+            listJsonLdEntity.addAll(getEntities(extractParam(it, q)))
+        }
+        return listJsonLdEntity
+    }
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt:70]: Ok I use this. I don't know `flatten`

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt [hunk 33]
```diff
@@ -0,0 +1,80 @@
...
+
+    suspend fun sendNotification(entity: JsonLdEntity, subscription: Subscription):
+        Triple<Subscription, Notification, Boolean> =
+        notificationService.callSubscriber(
+            subscription,
+            entity.id.toUri(),
+            entity
+        ).awaitFirst()
+
+    suspend fun getEntitiesToNotify(listEntityInfo: Set<EntityInfo>, q: String?): List<JsonLdEntity> {
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/job/TimeIntervalNotificationJob.kt:64]: ok I changed that

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt [hunk 34]
```diff
@@ -587,4 +586,36 @@ class SubscriptionService(
...
+                description, watched_attributes, time_interval, q,  notif_attributes, notif_format, endpoint_uri, 
+                endpoint_accept, endpoint_info,  status, times_sent, last_notification, last_failure, last_success, 
+                is_active, entity_info.id as entity_id, id_pattern, entity_info.type as entity_type, georel, geometry, 
+                coordinates, geoproperty
+            FROM subscription
+            LEFT JOIN entity_info ON entity_info.subscription_id = subscription.id
+            LEFT JOIN geometry_query ON geometry_query.subscription_id = subscription.id
+            WHERE time_interval IS NOT NULL
+            AND (last_notification IS NULL 
+            OR ((EXTRACT(EPOCH FROM last_notification) + time_interval) < EXTRACT(EPOCH FROM :currentDate)))
```
[the remark this answers is not in the document]
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/service/SubscriptionService.kt:603]: Like this is better 
```
AND (last_notification IS NULL 
                OR ((EXTRACT(EPOCH FROM last_notification) + time_interval) < EXTRACT(EPOCH FROM :currentDate))
)
```

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/WebClientConfig.kt (hunk 13 above)
Comment (GaelPoujolSenseen) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/config/WebClientConfig.kt:16]: They suggesting to use `AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager`

**Comment by bobeal** [comment]
Last points:
- Add the new environment variable in the provided docker-compose files in the project (only for unauthenticated mode)
- You'll have to rebase the branch on top of develop because there are some conflicts

**Comment by bobeal** [comment]
LGTM
